cmd/crawler/node/node.go

```diff
@@ -0,0 +1,433 @@
+	if len(discv5Nodes) == 0 && len(discvNodes) <= 0 {
+		log.Log.Warn("No bootstrap addresses supplied")
+	}
+	return discv5Nodes
+}
+
+func parseGenericAddrs(addrs []string) (qnodeString []string, multiAddrString []string) {
+	for _, addr := range addrs {
+		if addr == "" {
+			// Ignore empty entries
+			continue
+		}
+		_, err := qnode.Parse(qnode.ValidSchemes, addr)
+		if err == nil {
+			qnodeString = append(qnodeString, addr)
+			continue
+		}
+		_, err = multiAddrFromString(addr)
+		if err == nil {
+			multiAddrString = append(multiAddrString, addr)
+			continue
+		}
+		log.Log.Error(fmt.Sprintf("Invalid address of %s provided: %v", addr, err.Error()))
+	}
+	return qnodeString, multiAddrString
+}
+
+func multiAddrFromString(address string) (multiaddr.Multiaddr, error) {
+	addr, err := iaddr.ParseString(address)
+	if err != nil {
+		return nil, err
+	}
+	return addr.Multiaddr(), nil
+}
+
+func peersFromStringAddrs(addrs []string) ([]multiaddr.Multiaddr, error) {
+	var allAddrs []multiaddr.Multiaddr
+	qnodeString, multiAddrString := parseGenericAddrs(addrs)
+	for _, stringAddr := range multiAddrString {
+		addr, err := multiAddrFromString(stringAddr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Could not get multiaddr from string")
+		}
+		allAddrs = append(allAddrs, addr)
+	}
+	for _, stringAddr := range qnodeString {
+		qnodeAddr, err := qnode.Parse(qnode.ValidSchemes, stringAddr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Could not get qnode from string")
+		}
+		addr, err := convertToSingleMultiAddr(qnodeAddr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Could not get multiaddr")
+		}
+		allAddrs = append(allAddrs, addr)
+	}
+	return allAddrs, nil
+}
+
+func convertToSingleMultiAddr(node *qnode.Node) (multiaddr.Multiaddr, error) {
+	ip4 := node.IP().To4()
+	if ip4 == nil {
+		return nil, errors.Errorf("node doesn't have an ip4 address, it's stated IP is %s", node.IP().String())
+	}
+	pubkey := node.Pubkey()
+	assertedKey := convertToInterfacePubkey(pubkey)
+	id, err := peer.IDFromPublicKey(assertedKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get peer id")
+	}
+	multiAddrString := fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", ip4.String(), node.TCP(), id)
+	multiAddr, err := multiaddr.NewMultiaddr(multiAddrString)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get multiaddr")
+	}
+	return multiAddr, nil
+}
+
+func ConvertToInterfacePrivkey(privkey *ecdsa.PrivateKey) crypto.PrivKey {
+	typeAssertedKey := crypto.PrivKey((*crypto.Secp256k1PrivateKey)((*secp256k1.PrivateKey)(privkey)))
+	return typeAssertedKey
+}
+
+func convertToInterfacePubkey(pubkey *ecdsa.PublicKey) crypto.PubKey {
+	typeAssertedKey := crypto.PubKey((*crypto.Secp256k1PublicKey)((*secp256k1.PublicKey)(pubkey)))
+	return typeAssertedKey
+}
+
+func getConnPeerAddress(conn network.Conn) string {
+	return fmt.Sprintf("%s/ipfs/%s", conn.RemoteMultiaddr().String(), conn.RemotePeer())
```

Comment:
why ipfs?

Reply:
If I understand correctly, Multiaddr needs to be formed in this form, and ipfs is a fixed field

Reply:
https://github.com/multiformats/multiaddr has no requirement to fix to `ipfs` prefix, and we use `p2p` internally at https://github.com/Qitmeer/qitmeer/blob/0.10-dev/p2p/addrfactory.go#L134

Reply:
ok